sdks/java/core/src/main/java/org/apache/beam/sdk/io/TFRecordWriteSchemaTransformConfiguration.java

```diff
@@ -0,0 +1,118 @@
+
+import com.google.auto.value.AutoValue;
+import java.io.IOException;
+import org.apache.beam.sdk.io.fs.MatchResult;
+import org.apache.beam.sdk.schemas.AutoValueSchema;
+import org.apache.beam.sdk.schemas.annotations.DefaultSchema;
+import org.apache.beam.sdk.schemas.annotations.SchemaFieldDescription;
+import org.apache.beam.sdk.schemas.transforms.providers.ErrorHandling;
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.base.Strings;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+/**
+ * Configuration for reading from TFRecord.
+ *
+ * <p>This class is meant to be used with {@link TFRecordWriteSchemaTransformProvider}.
+ *
+ * <p><b>Internal only:</b> This class is actively being worked on, and it will likely change. We
+ * provide no backwards compatibility guarantees, and it should not be implemented outside the Beam
+ * repository.
+ */
+@DefaultSchema(AutoValueSchema.class)
+@AutoValue
+public abstract class TFRecordWriteSchemaTransformConfiguration {
+
+  public void validate() {
+    String invalidConfigMessage = "Invalid TFRecord Write configuration: ";
+
+    String filePath = getOutputPrefix();
+    try {
+      MatchResult matches = FileSystems.match(filePath);
+      checkState(!matches.metadata().isEmpty(), "Unable to find any files matching %s", filePath);
+    } catch (IOException e) {
+      throw new IllegalStateException(
+          String.format(invalidConfigMessage + "Failed to validate %s", filePath), e);
+    }
+
+    ErrorHandling errorHandling = getErrorHandling();
+    if (errorHandling != null) {
+      checkArgument(
+          !Strings.isNullOrEmpty(errorHandling.getOutput()),
+          invalidConfigMessage + "Output must not be empty if error handling specified.");
+    }
+  }
+
+  /** Instantiates a {@link TFRecordWriteSchemaTransformConfiguration.Builder} instance. */
+  public static TFRecordWriteSchemaTransformConfiguration.Builder builder() {
+    return new AutoValue_TFRecordWriteSchemaTransformConfiguration.Builder();
+  }
+
+  @SchemaFieldDescription("The directory to which files will be written.")
+  public abstract String getOutputPrefix();
+
+  @SchemaFieldDescription(
+      "The suffix of each file written, combined with prefix and shardTemplate.")
+  @Nullable
+  public abstract String getFilenameSuffix();
+
+  @SchemaFieldDescription("The number of shards to use, or 0 for automatic.")
+  public abstract int getNumShards();
+
+  @SchemaFieldDescription(
+      "The shard template of each file written, combined with prefix and suffix.")
+  @Nullable
+  public abstract String getShardTemplate();
+
+  @SchemaFieldDescription("Option to indicate the output sink's compression type. Default is NONE.")
+  public abstract String getCompression();
+
+  @SchemaFieldDescription(
+      "Whether to skip the spilling of data caused by having maxNumWritersPerBundle.")
+  public abstract boolean getNoSpilling();
```

Comment:
This (and other optional fields) should be marked nullable (e.g. https://github.com/apache/beam/blob/62152f404bf0877827fd0fb5ecc028fe58fade3d/sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaReadSchemaTransformConfiguration.java#L157). You will then also need to check if they are null in `expand`